### ISSUE TITLE
`DiscordGatewayClient` allocation and method improvements

### ISSUE THREAD

**Issue by VelvetThePanda** [Remora/Remora.Discord#254]
### Description

As it stands right now, the gateway client is not only the single biggest, single-purpose class* in the library, but it is also the top allocator as shown by a profiler. It's unlikely that these allocations pose a memory leak, but it is partially unnecessary when it's possible to reduce these allocations.

Allocations were initially my main goal when it came to sprucing up the client, as described below, but as I continued to look for room for improvement I realized just how much of a mess the client really is (and, I don't fault Jax here, as the client is somewhat non-trivial in its operation).

However, there is a lot of code that could be simplified by being broken out, as I'll describe later in this issue.

* This is excluding `____API` classes, which I would personally consider "multi-purpose". 

### Why This is Needed

As described above, the gateway client is responsible for the top five allocations across the entire span of the application lifetime, even in comparison to EF Core and the available Redis caching, which marshals objects when deserializing. 

The screenshot below is the measurement over a six-hour period, and while the overall allocations are rather trivial for most, and for the most part will be GC'd, these can still be potentially reduced by trivial means. 

![image](https://user-images.githubusercontent.com/42438262/202833530-77797d00-3405-4a40-b6b4-36a4b03753f7.png)

### Alternatives Considered

Short of rewriting the entire client (e.g. #176), there aren't many alternatives to be had here.

### Additional Details

Here is the meat of the document, and I'll point out everything that I think can be reasonably fixed. 

## Passing around `heartbeat_interval`
Call this a nitpick, but I feel like this issue is just large enough to be somewhat of an issue; not necessarily for maintainability or functionality, but for ease of writing code. There's also potential defensive copying at play here, though that's a bit of a micro-optimization.

https://github.com/Remora/Remora.Discord/blob/7935da197307034c5c8ffc8344cb766e40dcaa45/Backend/Remora.Discord.Gateway/DiscordGatewayClient.cs#L585

---

#### Heartbeating should be its own loop

As shown below, the gateway prioritizes heartbeats over user-submitted payloads which is *fine*, but I think this can be slightly improved by adjusting how payload rate-limiting is accounted for. 

Instead of having heartbeating and user payloads both be rate-limiting, we could instead reserve a single slot (heartbeating is ~45s, and it's 120 events/min, so there's only ever 1 heartbeat per ratelimit window).

That single slot ensures that heartbeats are always sent (and even if they aren't, this is technically OK because Discord doesn't immediately disconnect you from the socket). The only issue would be that the client automatically disconnects if the expected OP11 (`HEARTBEAT_ACK`) isn't received in time, but otherwise, this is a fine solution.

As for prioritization, a `heartbeatLock` field could be introduced; it'd be something along the lines of a `SemaphoreSlim` or even a `ReadWriteLockSlim`, and sending payloads waits on hearbeating to complete for consistency.

I'm pretty sure @carlst99 addressed this issue in the common gateway PR, but I'm not entirely certain.

https://github.com/Remora/Remora.Discord/blob/7935da197307034c5c8ffc8344cb766e40dcaa45/Backend/Remora.Discord.Gateway/DiscordGatewayClient.cs#L924-L963

#### Sending payloads shouldn't rely on sleep time

In addendum to the above, this code only exists because of the above and can be safely fixed with a concrete delay. For reasons I'll get into below, using `Task.Delay` isn't the best choice for this either.

https://github.com/Remora/Remora.Discord/blob/7935da197307034c5c8ffc8344cb766e40dcaa45/Backend/Remora.Discord.Gateway/DiscordGatewayClient.cs#L967-L972

https://github.com/Remora/Remora.Discord/blob/7935da197307034c5c8ffc8344cb766e40dcaa45/Backend/Remora.Discord.Gateway/DiscordGatewayClient.cs#L990-L1001

#### Stop using `Task.Delay` (Please)

This isn't a blanket statement; it's a suggestion to use the alternative, given it's appropriate for the context.

With Remora targeting .NET 6/7 now, we're given access to the new APIs introduced in those versions, which allow for simplifying code and reducing allocations with just a few lines of code being changed.

`PeriodicTimer` is a new timer class introduced in .NET 6, and is an alternative to previous timer APIs which were unideal for one reason or another. `PeriodicTimer`'s `WaitForNextTickAsync` method returns a `ValueTask<bool>`, which is already rather ideal. 

Its return is indicative of whether it will continue to tick, so much like a `ValueTask<Result>`, gateway termination can be gracefully handled. Unlike `Task.Delay`, cancelling the token passed to `WaitForNextTickAsync` simply returns `ValueTask<bool>(false)` while the former throws an exception which must be caught (and is typically measurably slower, though async tends to smooth these nuances).

This was what I initially set out to do, because one of, if not *the* top allocation throughout the lifetime of the gateway is `DelayPromise` and `TimerQueueTimer` ([`Task.Delay` allocates every time](https://source.dot.net/#System.Private.CoreLib/src/libraries/System.Private.CoreLib/src/System/Threading/Tasks/Task.cs,5618) you call it!). For this reason, it's *very* unideal to use `Task.Delay` in a tight/semi-tight loop, especially when its delay is static, as it'll cause lots of avoidable heap allocations.

There's two notable spots where this is an issue. 

https://github.com/Remora/Remora.Discord/blob/7935da197307034c5c8ffc8344cb766e40dcaa45/Backend/Remora.Discord.Gateway/DiscordGatewayClient.cs#L970

https://github.com/Remora/Remora.Discord/blob/7935da197307034c5c8ffc8344cb766e40dcaa45/Backend/Remora.Discord.Gateway/DiscordGatewayClient.cs#L632

---

# Nitpickings
<details><summary><h3><code>recommanded_shards</code> is not just a recommendation</h3></summary>

My issue with this section, the client treats `recommended_shards > set_shards` as a warning when the proper log level for this situation is `LogLevel.Error` (and potentially even returning a preemptive gateway error). The reason being is that if the bot is in > `2500 * shards` guilds, connecting to the gateway will ALWAYS return a 4010 (`INVALID_SHARD`) error. See [discord's docs](https://discord.com/developers/docs/topics/gateway#sharding)

https://github.com/Remora/Remora.Discord/blob/7935da197307034c5c8ffc8344cb766e40dcaa45/Backend/Remora.Discord.Gateway/DiscordGatewayClient.cs#L462-L473

</details>

<details><summary><h3>Re-calculating heartbeat windows</h3></summary>

This is a small issue; very small even, but it's something I picked up on regardless. Despite this only being done once per (re-)connection, I find it strange that this is recalculated. Then again shifting this up to the constructor could be considered bad practice, even if it's technically static-ish data.

https://github.com/Remora/Remora.Discord/blob/7935da197307034c5c8ffc8344cb766e40dcaa45/Backend/Remora.Discord.Gateway/DiscordGatewayClient.cs#L894-L898

</details>

<details><summary><h3><code>new</code>ing timeout policies all over the place</h3></summary>

I'm not entirely familiar with Polly, so I may be wrong here, but I find it somewhat daft that there are three instances of a timeout being `new`'d up, all with the same arbitrary(?) time limit, only to be used once. Are Polly timeout policies not thread-safe? Or I suppose not intended to be reused, otherwise, these should be merged into a field, IMO.

https://github.com/Remora/Remora.Discord/blob/7935da197307034c5c8ffc8344cb766e40dcaa45/Backend/Remora.Discord.Gateway/DiscordGatewayClient.cs#L545

https://github.com/Remora/Remora.Discord/blob/7935da197307034c5c8ffc8344cb766e40dcaa45/Backend/Remora.Discord.Gateway/DiscordGatewayClient.cs#L718

https://github.com/Remora/Remora.Discord/blob/7935da197307034c5c8ffc8344cb766e40dcaa45/Backend/Remora.Discord.Gateway/DiscordGatewayClient.cs#L545

</details>

<details><summary><h3>There should be a common `DisconnectAsync` method</h3></summary>

As it stands right now, there's no semblance of a disconnection method that handles the termination of the gateway (graceful or otherwise). Instead, inconsistent logic is sprinkled across call sites that varies based on context and can cause bugs.

https://github.com/Remora/Remora.Discord/blob/7935da197307034c5c8ffc8344cb766e40dcaa45/Backend/Remora.Discord.Gateway/DiscordGatewayClient.cs#L235-L250

https://github.com/Remora/Remora.Discord/blob/7935da197307034c5c8ffc8344cb766e40dcaa45/Backend/Remora.Discord.Gateway/DiscordGatewayClient.cs#L637-L664

</details>

**Comment by MazeXP** [comment]
A good writeup.

For the `PeriodicTimer` it would be important to provide a fallback for TFM netstandard2.1

**Comment by VelvetThePanda** [comment]
> A good writeup.
> 
> For the `PeriodicTimer` it would be important to provide a fallback for TFM netstandard2.1

I was under the impression that Remora would explicitly drop netstandard support in favor of .NET 6/7, since .NET Core 3.1 becomes EOL this year. If not, this would definitely need to be taken into consideration

**Comment by carlst99** [comment]
Nice writeup!

> I'm pretty sure @carlst99 addressed this issue in the common gateway PR, but I'm not entirely certain.
Yep, I added a 'first priority' send queue, which internal connection management packets were placed on. Anything on this queue was dispatched before any other enqueued packets and if I recall correctly, I ensured all priority packets were dispatched before allowing the send loop to check if it should sleep.

I was also toying around with a few allocation improvements in the transport service. I'd managed to cut back on send allocations significantly, not that those are particularly troublesome, and using the `RecyclableMemoryStream` for receiving data was looking promising.

**Comment by MazeXP** [comment]
> > A good writeup.
> > For the `PeriodicTimer` it would be important to provide a fallback for TFM netstandard2.1
> 
> Thanks! I was under the impression that Remora would explicitly drop netstandard support in favor of .NET 6/7, since .NET Core 3.1 becomes EOL this year. If not, this would definitely need to be taken into consideration.

NetStandard does not EOL with Core 3.1 cause it is still a thing with NET 7.0:
https://learn.microsoft.com/en-us/dotnet/standard/net-standard?tabs=net-standard-2-1#select-net-standard-version

But in the end it still depends if Remora wants to support all `netstandard` runtimes even if they are EOL or just those that are not EOL.

**Comment by VelvetThePanda** [comment]
I found it. [initial question](https://canary.discord.com/channels/789912892426027039/789913059116187658/955121817742360616) and his [initial answer](https://canary.discord.com/channels/789912892426027039/789913059116187658/955122278276943964), though everything's kind of up in the air since there hasn't been much further discussion since.

**Comment by Nihlus** [comment]
I have no intention of dropping support for .NET Standard - it will remain a runtime-agnostic library target.

**Comment by VelvetThePanda** [comment]
`#if` blocks it is, then.
